File: Binary search/python/74. Search a 2D Matrix.py
'''You are given an m x n integer matrix matrix with the following two properties:

Each row is sorted in non-decreasing order.
The first integer of each row is greater than the last integer of the previous row.
Given an integer target, return true if target is in matrix or false otherwise.

You must write a solution in O(log(m * n)) time complexity.

 

Example 1:


Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
Output: true
Example 2:


Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
Output: false
'''


class Solution:
    def searchMatrix(self, matrix, target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])

        left = 0 # start of the matrix
        right = rows * cols - 1 # end of the matrix

        while left <= right:
            mid = (left + right) // 2

            mid_val = matrix[mid // cols][mid % cols]  # 2D to 1D 

            # if we find the target, return true
            if mid_val == target:
                return True
            elif target > mid_val:
                # if the target is greater than the middle value, search the right half
                left = mid + 1
            else:
                # if the target is less than the middle value, search the left half
                right = mid - 1

        return False

 


if __name__ == "__main__":
    s = Solution()
    matrix = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
    target = 3
    print(s.searchMatrix(matrix, target))

File: Binary search/74. Search a 2D Matrix.js
`Write an efficient algorithm that searches for a value target in an m x n integer matrix matrix. This matrix has the following properties:

Integers in each row are sorted from left to right.
The first integer of each row is greater than the last integer of the previous row.
 

Example 1:


Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
Output: true
Example 2:


Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
Output: false`;

const searchMatrix = (matrix, target) => {
  let row = 0;
  let col = matrix[0].length - 1;

  while (row < matrix.length && col >= 0) {
    if (matrix[row][col] === target) return true;
    if (matrix[row][col] > target) {
      col--;
    } else {
      row++;
    }
  }
  return false;
};
console.log(
  searchMatrix(
    [
      [1, 3, 5, 7],
      [10, 11, 16, 20],
      [23, 30, 34, 60],
    ],
    3
  )
);

//  Binary search solution

const searchMatrix2 = (matrix, target) => {
  // get the rows and cols
  const rows = matrix.length;
  const cols = matrix[0].length;

  // left and right pointers for binary search
  let left = 0;
  let right = rows * cols - 1;

  while (left <= right) {
    // get the mid idx using Math.floor
    let mid = Math.floor((left + right) / 2);

    // get the midIdx element  Math.floor(mid/cols) => row, mid%cols => col
    const midElement = matrix[Math.floor(mid / cols)][mid % cols];

    if (midElement === target) {
      // if midElement === target, then return true
      return true;
    }

    if (midElement < target) {
      // if midElement < target, then we will search the right part of the matrix
      left = mid + 1;
    } else {
      // if midElement > target, then we will search the left part of the matrix
      right = mid - 1;
    }
  }

  return false;
};
